### PR TITLE
Add `unionWith` function to `balanced_mapTheory`

### PR DESCRIPTION
**Pull request by hrutvik**
As in Haskell's `Data.Map`, `unionWith` takes the union of two balanced maps, but uses its function parameter to combine values for keys present in both maps.

Theorems are stated in terms of a new constant in `finite_mapTheory`, `FMERGE_WITH_KEY` - this is like the existing `FMERGE`, but its function parameter processes the key as well as the values in the case where a key is present in both maps.

Some other theorems are also added to `finite_mapTheory` - these are mostly ported from https://github.com/cakeml/pure.

### CODE REVIEW

File: src/finite_maps/finite_mapScript.sml [hunk 1]
```diff
@@ -2943,6 +3129,91 @@ Proof
...
+
+Theorem FDIFF_FUPDATE:
+  FDIFF (fm |+ (k,v)) s =
+  if k IN s then FDIFF fm s else (FDIFF fm s) |+ (k,v)
+Proof
+  rw[fmap_eq_flookup, FLOOKUP_FDIFF, FLOOKUP_UPDATE] >>
+  EVERY_CASE_TAC >> gvs[]
+QED
+
+Theorem FDIFF_FEMPTY:
```
Comment (mn200) [inline src/finite_maps/finite_mapScript.sml:3187]: This should be `simp` (if something else doesn't already achieve that)

File: src/finite_maps/finite_mapScript.sml [hunk 2]
```diff
@@ -2943,6 +3129,91 @@ Proof
...
+  FDIFF FEMPTY s = FEMPTY
+Proof
+  rw[fmap_eq_flookup, FLOOKUP_FDIFF]
+QED
+
+Theorem FDIFF_EMPTY:
```
Comment (mn200) [inline src/finite_maps/finite_mapScript.sml:3193]: Also `simp`

File: src/finite_maps/finite_mapScript.sml [hunk 3]
```diff
@@ -2339,6 +2518,13 @@ val fmap_rel_FEMPTY2 = store_thm(
   METIS_TAC [FDOM_EQ_EMPTY]);
 val _ = export_rewrites ["fmap_rel_FEMPTY2"]
 
+Theorem fmap_rel_FEMPTY3:
```
Comment (mn200) [inline src/finite_maps/finite_mapScript.sml:2521]: This should be `simp`

File: src/finite_maps/finite_mapScript.sml [hunk 4]
```diff
@@ -1031,6 +1031,163 @@ val FLOOKUP_DRESTRICT = store_thm("FLOOKUP_DRESTRICT",
...
+
+Theorem FMERGE_WITH_KEY_DRESTRICT:
+  DRESTRICT (FMERGE_WITH_KEY f m1 m2) vs =
+  FMERGE_WITH_KEY f (DRESTRICT m1 vs) (DRESTRICT m2 vs)
+Proof
+  rw[fmap_eq_flookup, FLOOKUP_DRESTRICT, FLOOKUP_FMERGE_WITH_KEY] >>
+  every_case_tac >> gvs[]
+QED
+
+Theorem FMERGE_WITH_KEY_EQ_EMPTY:
```
Comment (mn200) [inline src/finite_maps/finite_mapScript.sml:1182]: This should be `simp`.

**Comment by mn200** [comment]
Looks good; would just add more `simp`s (as above)

**Comment by binghe** [comment]
Hi @hrutvik,

Nice work! May I suggest you to also add a little documentation for the existing `FMERGE` and your `FMERGE_WITH_KEY` into HOL Description (at the end of Section 5.5.5 "Finite maps (finite_map)", for example)?   I never knew this functionality of finite maps, as otherwise I may have already used it in the CCS example.

In case you refuse then I can try to do the documentation instead, but obviously you are the expert here.

Regards,

Chun Tian

**Comment by mn200** [comment]
I will merge this now, and await documentation as a separate PR.  Thanks for your work on this, Hrutvik!

**Pull request merged**